File: Code/Raspberry pi/xbee.py
#################################################################################
# Name: xbee.py                                                                 #
# Description: Xbee Communinication module                                      #
# Version: V1.1                                                                 #
# Author: Damilola Oludiran                                                     #
#                                                                               #
#                                                                               #
# This software is supplied "AS IS" without warranties of any kind.             #
#                                                                               #
#----------------------------------------------------------------------------   #
# History:                                                                      #
#            V1.0 Initial Version by Damilola Oludiran                          #
#            V1.1 Changed implementation to threaded class by Damilola Oludiran #
#                                                                               #
#################################################################################

#################################################################################
# Imports                                                                       #
#################################################################################
import serial
import threading
import queue as Queue
from time import sleep

#################################################################################
# Xbee Thread Class                                                             #
#################################################################################
class Xbee(threading.Thread):
    rxQueue = Queue.Queue(maxsize=1)
    stop = threading.Event()
    rx = True
    
    #################################################################################
    # Name: __init__                                                                #
    # Parameters: self                                                              #
    # Description: This method initialises the thread and serial interface          #
    #                with xbee                                                      #
    # Preconditions: The xbee circuit is connected                                  #
    # Postconditions: Thread starts, xbee is initialized                            #
    #################################################################################
    def __init__(self, port = "/dev/ttyS0", baudrate=9600):
        threading.Thread.__init__(self)
        self.serial = serial.Serial(port = port, baudrate=baudrate)
        to_send1 =[0x42]
        self.serial.write(serial.to_bytes(to_send1))
        self.start()
    
    #################################################################################
    # Name: shutdown                                                                #
    # Parameters: self                                                              #
    # Description: This method terminates the thread                                #
    # Preconditions:n/a                                                             #
    # Postconditions: Thread stops running                                          #
    #################################################################################
    def shutdown(self):
        self.stop.set()
        self.join()
    
    #################################################################################
    # Name: run                                                                     #
    # Parameters: self                                                              #
    # Description: This method is the thread's main loop                            #
    # Preconditions: n/a                                                            #
    # Postconditions: Continuously retrieves xbee data                              #
    #################################################################################
    def run(self):
        while not self.stop.is_set():
            self.Rx()
            
            

    #################################################################################
    # Name: recieve                                                                 #
    # Parameters: self                                                              #
    # Description: This method returns the current and voltage data                 #
    # Preconditions: rxQueue is initialised                                         #
    # Postconditions: returns queue data if any                                     #
    #################################################################################
    def recieve(self):
        if not self.rxQueue.empty():
            return self.rxQueue.get()
        else:
            return 0

    #################################################################################
    # Name: send                                                                    #
    # Parameters: self                                                              #
    # Description: This method sends data over serial                               #
    #                                                                               #
    # Preconditions: xbee is initialised                                            #
    # Postconditions: data is sent                                                  #
    #################################################################################
    def send(self, data, address=0xFFFF, frameid = 0x00, options=0x00):
        if not data:
            return 0
        message = '7E 00 {:02X} 01 {:02X} {:02X} {:02X} {:02X}'.format(
            len(data) + 5, frameid, (address & 0xFF00) >> 8, address & 0xFF, options
            )

        frame = bytearray.fromhex(message)
        frame.extend(data)

        frame.append(0xFF - sum(frame[3:]) & 0xFF)

        return self.serial.write(frame)

    #################################################################################
    # Name: Rx                                                                      #
    # Parameters: self                                                              #
    # Description: This method recieves data from serial interface                  #
    # Preconditions: xbee is initialised                                            #
    # Postconditions: data is recieved, if any                                      #
    #################################################################################
    def Rx(self):
        message = []
        i = 0
        start = self.serial.read() 
        if (start.hex() == '7e'):
            message = self.serial.read(12)
            print(message.hex())
            if (message[4] == 0xde):
                current = ((message[7] & 0x00FF) << 8) + message[8]
                voltage = ((message[9] & 0x00FF) << 8) + message[10]
                data =[current, voltage]
                self.rxQueue.put(data)

File: Code/PowerMonitoring/xbee.c
/**
// Code provided by Damilola for system integration
 ******************************************************************************
 * Name:    xbee.c
 * Description: XBee interface Configuration Implementation
 * Version: V1.0
 *  Author: Damilola Oludiran 
  
 *
 * This software is supplied "AS IS" without warranties of any kind.
 *
 *
 *----------------------------------------------------------------------------
 * History:
 *          V1.0 Initial Version by Damilola Oludiran
 *            
 *         
 *****************************************************************************
 */

#include"xbee.h"
/**
* Name:           void xbee_init()
* Paramaters:     none
* Description:    This function configures USART2 and xbee
* Preconditions:  stm32f10x.h is included in project
* Postconditions: 
*/

void xbee_init(void)
{
	usart_init();
	usart_tx(0x42);
}


/**
* Name:           xbee_send
* Paramaters:     uint16_t address, uint16_t data
* Description:    This function sends data to the xbee module
* Preconditions:  16 bit Destination Address and 16 bit Data
* Postconditions: 
*/
void xbee_send(uint16_t address, uint16_t current, uint16_t voltage)
{
	const uint16_t message_buff_size = 13;
	uint16_t message[message_buff_size];
	uint8_t checksum = 0;
	int i = 0;
	
	message[0] = 0x7E;
	message[1] = 0x00;
	message[2] = 0x09;
	message[3] = 0x01;
	message[4] = 0x00;
	message[5] = (address & 0xFF00) >> 8;
	message[6] = (address & 0xFF);
	message[7] = 0x00;
	message[8] = (current & 0xFF00) >> 8;
	message[9] = (current & 0xFF);
	message[10] = (voltage & 0xFF00) >> 8;
	message[11] = (voltage & 0xFF);
	
	for ( i = 3; i < message_buff_size - 1; i++)
	{checksum += message[i];}
	
	checksum = 0xFF -(checksum & 0xFF);
	
	message[12] = checksum;
	
	
	for ( i = 0; i < message_buff_size; i++)
	{usart_tx(message[i]);}
	
}

/**
* Name:           xbee_recieve
* Paramaters:     n/a
* Description:    This function retrieves data from the xbee module
* Preconditions:  n/a
* Postconditions: 16 bit data output
*/

uint16_t xbee_recieve(void)
{
	uint16_t data;
	uint16_t dataMSB;
	uint16_t dataLSB;

 if ((GLOBAL_DATA_RECEIVED == true) && (GLOBAL_USART_Received_Data[0] == 0x7E))
 {
	 dataMSB = (GLOBAL_USART_Received_Data[8] & 0xFF) << 8;
	 dataLSB = (GLOBAL_USART_Received_Data[9] & 0xFF);
	 data = dataMSB + dataLSB;
	 
	 GLOBAL_DATA_RECEIVED = false;
	 return data;
 }
 else{return 0;}
	
	
}

